Move Deployments to ReplicaSets and switch the Deployment selector to the new LabelSelector.

madhusudancs opened the pull request:
**Release note for v1.2** (Copying @bgrant0607's comment from below for better visibility)

Delete all Deployments, their RCs, and their pods, and re-create new ones using selector.matchLabels.

Way to not disrupt the running pods:
1. Do a non-cascading delete of the Deployment and its RCs
2. Upgrade Kubernetes.
3. Create ReplicaSets corresponding to the RCs using selector.matchLabels, which will adopt the pods
4. Re-create the Deployment using selector.matchLabels

**End of release note**

@bgrant0607 I have added the "release-note-breaking-change" label, but I am not sure if this change qualifies for a release note describing "a way to achieve or approximate previous behavior, or migration that should/could be done prior to upgrade" that you mentioned in your email. Do you think this deserves a description?

Update the Deployments' API types, defaulting code, conversions, helpers
and validation to use ReplicaSets instead of ReplicationControllers and
LabelSelector instead of map[string]string for selectors.

Also update the Deployment controller, registry, kubectl subcommands,
client listers package and e2e tests to use ReplicaSets and
LabelSelector for Deployments.

Only the last two commits in this PR need reviews.

madhusudancs committed:
Move Deployments to ReplicaSets and switch the Deployment selector to the new LabelSelector.

Update the Deployments' API types, defaulting code, conversions, helpers
and validation to use ReplicaSets instead of ReplicationControllers and
LabelSelector instead of map[string]string for selectors.

Also update the Deployment controller, registry, kubectl subcommands,
client listers package and e2e tests to use ReplicaSets and
LabelSelector for Deployments.

bgrant0607 commented:
Yes, we should describe how to migrate:

Simple way:

Delete all Deployments, their RCs, and their pods, and re-create new ones using selector.matchLabels.

Way to not disrupt the running pods:

1. Do a non-cascading delete of the Deployment and its RCs
2. Upgrade Kubernetes.
3. Create ReplicaSets corresponding to the RCs using selector.matchLabels, which will adopt the pods
4. Re-create the Deployment using selector.matchLabels

bgrant0607 commented:
Could one of @janetkuo, @nikhiljindal, @kargakis, or @ironcladlou please review this?

kargakis commented:
This PR is insanely huge. I would propose breaking it down to smaller self-contained PRs (1. addition of Replica Sets, 2. Move Deployments to use Replica Sets, 3. Change Deployment label selector)

kargakis commented:
Ah I now noticed that only the two last commits need reviewing here. Sorry for the noise

bgrant0607 commented:
@kargakis Yes, all the ReplicaSet PRs seemed to be blocked on the first PR in the submit queue.

madhusudancs commented:
> This PR is insanely huge. I would propose breaking it down to smaller self-contained PRs (1. addition of Replica Sets, 2. Move Deployments to use Replica Sets, 3. Change Deployment label selector)

Also just to clarify, it doesn't make sense to move deployments to replica sets and then switch to label selector in two separate PRs while having the build and all the tests pass in both the PRs.

bgrant0607 commented:
LGTM once comments from previous PRs are addressed. 

Will wait for previous PRs to merge before applying LGTM label. This will need to be rebased, regenerate code, etc., anyway.

madhusudancs committed:
Make deployments work.

madhusudancs committed:
Auto-generated code.

bgrant0607 commented:
This has to be hard to keep up to date.

Tests are complaining about a flag:
```
Verifying ./hack/../hack/verify-flags-underscore.py
Found a flag declared with an _ but which is not explicitly listed as a valid flag name in hack/verify-flags/excluded-flags.txt
Are you certain this flag should not have been declared with an - instead?
concurrent_rs_syncs
FAILED
```
Should be concurrent-replicaset-syncs

janetkuo commented:
@madhusudancs I'd like to inform you that all deployment e2e tests are removed from build test (see #19667 and https://github.com/kubernetes/kubernetes/blob/master/docs/devel/e2e-tests.md#kinds-of-tests) so that they won't block merge queue. Therefore when you make code changes to deployment related stuffs please test all deployment e2e tests locally and make sure they all passed. Without this, any PR shouldn't be merged. Thanks!

madhusudancs commented:
@janetkuo Oh, thanks for pointing that out. I was happily assuming that all the tests were passing. Will test locally.

bgrant0607 commented:
Tests failed:
```
'kubectl scale --current-replicas=3 --replicas=1 deployment/nginx-deployment' exited with status 1
```

smarterclayton commented:
Quick question - why aren't we supporting ReplicaSet -> ReplicationControllers migration in a backward compatible fashion?  Why aren't they both stored in etcd under the same key, and both the old and new paths resources them?  Because of the label selector?

madhusudancs commented:
@janetkuo @bgrant0607 PTAL.

@janetkuo Also, thanks for that pointer again. That uncovered two e2e test failures. I have fixed both of them here in this PR.

@smarterclayton yes, because of the new label selector semantics. That was the whole reason we switched Deployments from ReplicationController -> ReplicaSet.

janetkuo commented:
There should be implementation for replicasets printer/describer somewhere, if not in this PR. 
Refer to:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L1438-L1467
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/describe.go#L1600-L1641

janetkuo commented on pkg/apis/extensions/helpers.go:
```diff
@@ -16,24 +16,32 @@ limitations under the License.
-import (
-	"k8s.io/kubernetes/pkg/api"
-)
+// TODO(madhusudancs): Fix this when Scale group issues are resolved.
```
Put the issue / PR number here for easier tracking?

madhusudancs replied:
Done.

janetkuo commented on hack/test-cmd.sh:
```diff
@@ -963,16 +964,16 @@ __EOF__
-  kubectl delete deployment/nginx-deployment "${kube_flags[@]}"
-  # TODO: Remove once deployment reaping is implemented
-  kubectl delete rc --all "${kube_flags[@]}"
+  # ### Scale a deployment
```
Add TODO here for scale too?

bgrant0607 replied:
Why is this commented out? Because reaping support isn't yet merged?

madhusudancs replied:
@bgrant0607 No, no. Because Scale is disabled for Deployments for now.

madhusudancs replied:
@janetkuo Done.

janetkuo commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -220,7 +220,7 @@ type DeploymentSpec struct {
 
 	// Label selector for pods. Existing ReplicationControllers whose pods are
 	// selected by this will be the ones affected by this deployment.
-	Selector map[string]string `json:"selector,omitempty"`
```
ReplicationControllers --> ReplicaSets

bgrant0607 commented:
Good catch, @janetkuo. 

@madhusudancs Please implement get/describe support for ReplicaSet in another PR.

janetkuo commented on pkg/kubectl/describe.go:
```diff
@@ -1612,7 +1612,7 @@ func (dd *DeploymentDescriber) Describe(namespace, name string) (string, error)
 		fmt.Fprintf(out, "CreationTimestamp:\t%s\n", d.CreationTimestamp.Time.Format(time.RFC1123Z))
 		fmt.Fprintf(out, "Labels:\t%s\n", labels.FormatLabels(d.Labels))
-		fmt.Fprintf(out, "Selector:\t%s\n", labels.FormatLabels(d.Spec.Selector))
+		fmt.Fprintf(out, "Selector:\t%s\n", d.Spec.Selector)
```
Use 	`selector, err := unversioned.LabelSelectorAsSelector(d.Spec.Selector)` instead and print that returned `selector`.

madhusudancs replied:
Done.

madhusudancs committed:
Address review comments.

madhusudancs commented:
@bgrant0607 @janetkuo Ack. I will implement get/describe support for ReplicaSet in a follow up PR.

janetkuo commented on pkg/apis/extensions/v1beta1/types.go:
```diff
@@ -233,7 +233,7 @@ type DeploymentSpec struct {
 	MinReadySeconds int32 `json:"minReadySeconds,omitempty"`
 
-	// The number of old ReplicationControllers to retain to allow rollback.
+	// The number of old ReplicaSets to retain to allow rollback.
```
We need this change in `pkg/apis/extensions/types.go` too

madhusudancs replied:
Done.

bgrant0607 commented:
Note that this will need to be rebased after #20368 is merged (assuming it merges first).

janetkuo commented:
Please also add replicasets to the list of valid resources in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cmd.go#L113

`pkg/apis/extensions/v1beta1/types_swagger_doc_generated.go` should be updated to reflect changes in types.go, but the check passed. Looks like `hack/verify-generated-swagger-docs.sh` is missing in travis. cc @ixdy

janetkuo commented on pkg/apis/extensions/validation/validation.go:
```diff
@@ -346,9 +346,24 @@ func ValidateRollback(rollback *extensions.RollbackConfig, fldPath *field.Path)
-	allErrs = append(allErrs, apivalidation.ValidatePodTemplateSpecForRC(&spec.Template, spec.Selector, spec.Replicas, fldPath.Child("template"))...)
+
+	if spec.Selector == nil {
+		allErrs = append(allErrs, field.Required(fldPath.Child("selector"), ""))
```
Isn't `spec.Selector` not required?

madhusudancs replied:
IIUC, selector is optional only in the external version. By the time it is converted to an internal object, it is defaulted to `Spec.Template.Labels`. Since the validation happens on internal objects, validation should fail if it is nil, IMHO.

The second thing is, this code is trying to preserve the semantics of what RC based `ValidateDeploymentSpec()` did. Before, this function would call `ValidateNonEmptySelector()` and and that would call 

```
selector := labels.Set(selectorMap).AsSelector()
if selector.Empty() {
	allErrs = append(allErrs, field.Required(fldPath, ""))
}
```

And because of the way this worked, selector.Empty() would return true for nil SelectorMap and then there would be error. I am trying to preserve that behavior here.

janetkuo replied:
I see. Thanks for the explanation.

janetkuo commented on pkg/apis/extensions/validation/validation.go:
```diff
@@ -346,9 +346,24 @@ func ValidateRollback(rollback *extensions.RollbackConfig, fldPath *field.Path)
+
+	selector, err := unversioned.LabelSelectorAsSelector(spec.Selector)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "failed to convert LabelSelector to Selector."))
```
This error message may confuse the users... the conversion between LabelSelector and Selector is implementation detail; the users only know that they have `selector` in deployment.spec.

madhusudancs replied:
After reading your comment, I think I agree. But that also means we need to change this in a lot of other places as well. Would you be Ok if these changes are in a separate follow up PR?

janetkuo replied:
@madhusudancs sounds good

madhusudancs commented:
@janetkuo PTAL.

Ran the e2e tests manually again with --test_args='--ginkgo.focus=Deployment.*' and they all passed.

janetkuo commented on pkg/kubectl/cmd/expose.go:
```diff
@@ -149,9 +149,9 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 	// For objects that need a pod selector, derive it from the exposed object in case a user
 	// didn't explicitly specify one via --selector
 	if s, found := params["selector"]; found && kubectl.IsZero(s) {
-		s, err := f.PodSelectorForObject(inputObject)
```
Will expose support replicasets? If so, we need to add replicasets to `CanBeExposed` in factory.go. And modify the description of `kubectl expose` in this file too. This can be done in a follow up PR.

madhusudancs replied:
@janetkuo yes, added the support in this PR - https://github.com/kubernetes/kubernetes/pull/20928. PTAL.

janetkuo commented on pkg/kubectl/cmd/util/factory.go:
```diff
@@ -257,7 +261,41 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
+					return "", fmt.Errorf("the service has no pod selector set")
+				}
+				return kubectl.MakeLabels(t.Spec.Selector), nil
+			case *extensions.Deployment:
```
Can you explain the differences between `MapBasedSelectorForObject` and `PodSelectorForObject` (why do we need to separate them)? Is it because `ParseLabels` in pkg/kubectl/generate.go can only recognize and convert labels/selectors like `app=nginx` to `map[string]string` but not those like `app in (nginx)`?

madhusudancs replied:
Yes exactly, that's the reason. And did not want to change the signature of that function.

janetkuo replied:
If so do we need to support deployment for `PodSelectorForObject`? Do we need to support rc, pod, service for `MapBasedSelectorForObject`?

madhusudancs replied:
The idea is that, PodSelectorForObject() returns the stringified/serialized version of the pod selector no matter which object you give, provided the object has object.Spec.Selector field (that reminds me I should add ReplicaSet to this function as well in my other PR). MapBasedSelectorForObject() returns only the map[string]string component. And both of them do that consistently for all the objects.if

janetkuo replied:
OK

janetkuo commented on pkg/kubectl/history.go:
```diff
@@ -67,11 +67,11 @@ func (h *DeploymentHistoryViewer) History(namespace, name string) (HistoryInfo,
 		return historyInfo, fmt.Errorf("failed to retrieve deployment %s: %v", name, err)
 	}
-	_, allOldRCs, err := deploymentutil.GetOldRCs(*deployment, h.c)
+	_, allOldRCs, err := deploymentutil.GetOldReplicaSets(*deployment, h.c)
```
`allOldRSs` and `newRS` (below)?

janetkuo replied:
Also there's `allRCs` etc

madhusudancs replied:
Fixed.

janetkuo commented on pkg/util/deployment/deployment.go:
```diff
@@ -39,112 +40,124 @@ const (
-func GetNewRCFromList(deployment extensions.Deployment, c clientset.Interface, getRcList func(string, api.ListOptions) ([]api.ReplicationController, error)) (*api.ReplicationController, error) {
+// GetNewReplicaSetFromList returns a replica set that matches the intent of the given deployment; get ReplicaSetList with the input function.
+// Returns nil if the new replica set doesnt exist yet.
+func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Interface, getRcList func(string, api.ListOptions) ([]extensions.ReplicaSet, error)) (*extensions.ReplicaSet, error) {
```
Rename `getRcList` to `getRsList` (or `getRSList`) everywhere in this file.

madhusudancs replied:
Done.

janetkuo commented on pkg/util/labels/labels.go:
```diff
@@ -51,3 +53,45 @@ func CloneAndRemoveLabel(labels map[string]string, labelKey string) map[string]s
+			newMExps[i].Operator = me.Operator
+			if me.Values != nil {
+				newMExps[i].Values = make([]string, len(me.Values))
+				for j, val := range me.Values {
```
Use built-in `copy`?

madhusudancs replied:
I think that only copies slices. We could probably make a slice containing the map and copy them, but that's not worth it.

janetkuo replied:
I meant `copy(newMExps[i].Values, me.Values)`.

madhusudancs replied:
Ah Ok. Done. PTAL.

janetkuo commented:
@madhusudancs LGTM other than a minor suggestion. Feel free to add the tag.

madhusudancs commented:
@janetkuo Thanks! Addressed the comment. 

Adding LGTM.

kargakis commented on pkg/kubectl/cmd/expose.go:
```diff
@@ -149,9 +149,9 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 	// didn't explicitly specify one via --selector
 	if s, found := params["selector"]; found && kubectl.IsZero(s) {
-		s, err := f.PodSelectorForObject(inputObject)
+		s, err := f.MapBasedSelectorForObject(inputObject)
```
What's the difference between `MapBasedSelectorForObject` and `PodSelectorForObject`? Do we need both?

kargakis replied:
Just saw the discussion in https://github.com/kubernetes/kubernetes/pull/19840/files#r52379928

@madhusudancs I think you can safely collapse both methods into one as expose is the only place they are used.

madhusudancs replied:
@kargakis Since this is in pkg/kubectl/... and an exported function, we weren't sure who else outside Kubernetes was using this, particularly OpenStack. So we left the existing function as is. If you are certain that OpenStack doesn't use the factory, we can probably eliminate one of the two functions.

cc @bgrant0607

kargakis replied:
We are using the old method but the signatures are the same and we can repurpose factory methods however we like. That being said, let's keep it around and once we rebase and make use of the new method then I'll submit a fix to remove the old one.

madhusudancs replied:
Sure, thanks. We did not want to surprise.

kargakis commented on pkg/kubectl/scale.go:
```diff
@@ -327,53 +328,57 @@ func (precondition *ScalePrecondition) ValidateDeployment(deployment *extensions
-	}
-	return nil
-}
+// TODO(madhusudancs): Fix this when Scale group issues are resolved (see issue #18528).
```
@bgrant0607 why the current state of the scale subresource warrants disabling of deployment scaling?

bgrant0607 replied:
@kargakis The scale subresource has been temporarily removed from Deployment, until the new autoscaling API group is created and we figure out how to register an endpoint for a resource from a different API group from the parent resource.
